mono/metadata/runtime.c

```diff
@@ -83,6 +92,38 @@ mono_runtime_fire_process_exit_event (void)
 #endif
 }
 
+static void
+mono_runtime_call_atexit (void)
+{
+	MonoAtExitData *at;
+	for (at = mono_atexit; at; at = at->next)
+		at->cb (at->user_data);
+}
+
+/**
+ * mono_runtime_atexit:
+ *
+ * Register a callback to be invoked when the runtime begins to shutdown.
+ *
+ * This will be invoked both for active (calling Environment::Exit) and passive shutdown (all threads gone).
+ *
+ * The runtime will be functional at the time the callback is invoked but managed code will already have begun
+ * to cleanup. This means the callback must not block or depend on execution of managed code.
+ *
+ */
+void
+mono_runtime_atexit (MonoAtExitCallback callback, void *user_data)
```

Comment:
The function should check whether `callback` is NULL before registering an entry for it, or otherwise `mono_runtime_call_atexit` will bomb